README
--------------------------------------------------------------------------
XSIM     2022/02/12 v0.21a
XSIMVERB 2022/02/12 v0.4

  eXercise Sheets IMproved

--------------------------------------------------------------------------
Clemens Niederberger
Web:    https://github.com/cgnieder/xsim
E-Mail: clemens@cnltx.de
--------------------------------------------------------------------------
Copyright 2017--2022 Clemens Niederberger

This work may be distributed and/or modified under the
conditions of the LaTeX Project Public License, either version 1.3c
of this license or (at your option) any later version.
The latest version of this license is in
  http://www.latex-project.org/lppl.txt
and version 1.3c or later is part of all distributions of LaTeX
version 2008/05/04 or later.

This work has the LPPL maintenance status `maintained'.

The Current Maintainer of this work is Clemens Niederberger.
--------------------------------------------------------------------------
The xsim package consists of the code files:
   xsim.sty
   xsimverb.sty
   xsim.style.layouts.code.tex
--------------------------------------------------------------------------
The xsim package consists of the documentation files:
   README, HISTORY
   xsim-manual.tex, xsim-manual.pdf
   xsim-manual.cls
--------------------------------------------------------------------------
The xsim package consists of the example files:
   xsim.blanks.tex, xsim.blanks.pdf
   xsim.boxed-headings.tex, xsim.boxed-headings.pdf
   xsim.code-and-output.tex, xsim.code-and-output.pdf
   xsim.collections.tex, xsim.collections.pdf
   xsim.crossref.tex, xsim.crossref.pdf
   xsim.description-list.tex, xsim.description-list.pdf
   xsim.different-point-types.tex, xsim.different-point-types.pdf
   xsim.difficulties.tex, xsim.difficulties.pdf
   xsim.floating.tex, xsim.floating.pdf
   xsim.grade-distribution.tex, xsim.grade-distribution.pdf
   xsim.hints.tex, xsim.hints.pdf
   xsim.listings.tex, xsim.listings.pdf
   xsim.listofexercises.tex, xsim.listofexercises.pdf
   xsim.multiplechoice.tex, xsim.multiplechoice.pdf
   xsim.pointsums.tex, xsim.pointsums.pdf
   xsim.randomexercises.tex, xsim.randomexercises.pdf
   xsim.various.tex, xsim.various.pdf
   xsim.texsx-13635.tex, xsim.texsx-13635.pdf
   xsim.texsx-155630.tex, xsim.texsx-155630.pdf
   xsim.texsx-199360.tex, xsim.texsx-199360.pdf
   xsim.texsx-299534.tex, xsim.texsx-299534.pdf
   xsim.texsx-305110.tex, xsim.texsx-305110.pdf
   xsim.texsx-308883.tex, xsim.texsx-308883.pdf
   xsim.texsx-338165.tex, xsim.texsx-338165.pdf
   xsim.texsx-350028.tex, xsim.texsx-350028.pdf
   xsim.texsx-369065.tex, xsim.texsx-369065.pdf
   xsim.texsx-369803.tex, xsim.texsx-369803.pdf
   xsim.texsx-369686.tex, xsim.texsx-369686.pdf
   xsim.texsx-370642.tex, xsim.texsx-370642.pdf
   xsim.texsx-391530.tex, xsim.texsx-391530.pdf
   xsim.texsx-395273.tex, xsim.texsx-395273.pdf
   xsim.texsx-466584.tex, xsim.texsx-466584.pdf
   xsim.texsx-498299.tex, xsim.texsx-498299.pdf
   xsim.texsx-549540.tex, xsim.texsx-549540.pdf
   xsim.texsx-576998.tex, xsim.texsx-576998.pdf
   xsim.texsx-580636.tex, xsim.texsx-580636.pdf
   xsim.texwelt-6698.tex, xsim.texwelt-6698.pdf
   xsim.texwelt-15093.tex, xsim.texwelt-15093.pdf
   xsim.texwelt-23968.tex, xsim.texwelt-23968.pdf
   xsim.golatex-80640.tex, xsim.golatex-80640.pdf
   xsim.golatex-91339.tex, xsim.golatex-91339.pdf
   xsim.issues-49.tex, xsim.issues-49.pdf
--------------------------------------------------------------------------
If you have any ideas, questions, suggestions or bugs to report, please
feel free to contact me.
--------------------------------------------------------------------------


HISTORY
--------------------------------------------------------------------------

XSIM -- eXercise Sheets IMproved

--------------------------------------------------------------------------
Clemens Niederberger
Web:    https://github.com/cgnieder/xsim
E-Mail: clemens@cnltx.de
--------------------------------------------------------------------------
Copyright 2017--2022 Clemens Niederberger

This work may be distributed and/or modified under the
conditions of the LaTeX Project Public License, either version 1.3c
of this license or (at your option) any later version.
The latest version of this license is in
  http://www.latex-project.org/lppl.txt
and version 1.3c or later is part of all distributions of LaTeX
version 2008/05/04 or later.

This work has the LPPL maintenance status `maintained'.

The Current Maintainer of this work is Clemens Niederberger.
--------------------------------------------------------------------------

HISTORY:

 v0.0   2017/02/09 - start with coding the first ideas
 v0.1   2017/03/19 - first draft which actually seems to do everything as
                     planned
 v0.2   2017/03/21 - bugfix in the verbwrite module
                   - add `use!' and `print!' properties
                   - add the concepts of `noupdate' and `boolean' properties
 v0.4   2017/03/24 - \IfExerciseBooleanPropertyTF
                   - \GetExercisePropertyTF, \PropertyValue
                   - \printallsolutions
                   - \UseExerciseTemplate
                   - bug fix in solution environments: now get printed again
                     when option `print=true' is used
                   - bug fix to the translations of the solution list headings:
                     exercise name instead of exercise type is now printed
                   - \AddtoExerciseTypeGoal, \AddtoExerciseGoal
                     \AddtoExerciseTypeGoalPrint, \AddtoExerciseGoalPrint,
                     \ExerciseGoalValuePrint
                   - \addpoints, \addbonus, \points, \printgoal
 v0.5   2017/03/30 - add style file mechanism
                   - make solution print conditions `section' and `chapter'
                     combinable
 v0.5a  2017/04/15 - fix wrong translation of the `default-heading'
 v0.5b  2017/04/16 - add \SetExerciseProperty
 v0.6   2017/04/27 - official interface for grade distributions
                   - refine hook management (environment module)
 v0.6a  2017/04/28 - refine \blank command, add documentation for said command
 v0.7   2017/05/09 - \xsim_list_length:n
                   - new boolean property `used'
                   - when an exercise is collected the counter isn't stepped
                     but only when the exercise is inserted for the first
                     time; when this happens the new property `used' is set to
                     `true' 
                   - make setting boolean properties case insensitive
                   - add missing pieces to manual (eg `page' property...)
                   - \TotalExerciseGoals and \TotalExerciseTypeGoals
                   - \XSIMifchapterTF
                   - \xsim_if_goal_sum:nnnTF and friends,
                     \IfExerciseTypeGoalsSumTF and \IfExerciseGoalsSumTF
                   - some more hints in the manual regarding collections
 v0.8   2017/05/18 - bug fix in \xsim_set_property:nnnn
                   - random exercise lists from collections
                   - \xsim_collection_count:n
                   - \GetExerciseIdForProperty and \GetExerciseTypeForProperty
                     for unique properties
                   - set property `print=false' during collection and only to
                     true when the corresponding exercises actually are
                     printed
                   - do not insert \par at the beginning and end of exercises
                     and solutions; this should be left to the templates!
                   - new option `print-collection/print' which allows to
                     choose between exercises, solutions, and both to be
                     printed
                   - new parameter `solution-counter' in order to avoid
                     possible conflicts with existing counters
 v0.8a  2017/05/19 - \IfPrintSolutionTF
 v0.8b  2017/05/23 - use \printgoal in default environment template
                   - \xsim_print_goal:nnn now expands input to a number
 v0.9   2017/06/20 - \xsimstyle's behaviour now matches the description in
                     the manual
                   - add style file `layouts' which at the moment provides the
                     two exercise templates `runin' and `margin'
                   - \printforexercises , \xsim_print_type_code:nnnnn
                   - \IfExerciseParameterSetTF
                   - \GetExerciseParameterTF
                   - change syntax of \SetExerciseProperty
                   - \SetExpandedExerciseProperty
                   - \ExerciseSetProperty, \ExerciseSetExpandedProperty
                   - property `sectioning'
                   - \IfExistSolutionTF
 v0.10  2017/09/19 - \xsim_input:nnn
                   - \GetExerciseBody
                   - option `blank/fill'
                   - bug fix in \xsim_start_collection:nn
                   - added further examples to package
                   - new option `file-extension'
 v0.10a 2017/10/01 - bug fix: grading table options can now be set with
                     \xsimsetup (issue 15)
 v0.11  2018/02/12 - change name requirement for style files in order to
                     prevent possible name clashes with module files
                   - fix bug in \SetExerciseProperty and
                     \SetExpandedExerciseProperty
                   - fix bug when xsimverb is used standalone (issue 21)
                   - add option `split-aux-lists' (issue 18)
                   - \IfExerciseTagSetTF and \IfExerciseTopicSetTF
                   - introduce a \xsim_if_<tagging>_value:nTF for each tag

build.lua
-- Build script for xsim
module  = "xsim"

xsimpkgv  = "0.19a"
xsimpkgd  = "2020/03/16"
xsimverbv = "0.1a"
xsimverbd = "2019/12/29"

docfiledir    = "doc"
sourcefiledir = "code"
textfiles     = {"./README"}
docfiles      = {"*.history", "*.ideas", "*.tex","examples/*.*", "*.cls"}
sourcefiles   = {"*.tex", "*.sty"}
installfiles  = {"code/*.*", "doc/*.*","doc/examples/*.pdf", "README"}
flatten       = false
packtdszip    = false
cleanfiles    = {}
tdslocations  = {
  "doc/latex/xsim/xsim.history",
  "doc/latex/xsim/xsim.ideas",
  "doc/latex/xsim/xsim_manual.pdf",
  "doc/latex/xsim/xsim_manual.tex",
  "doc/latex/xsim/README",
  "doc/latex/xsim/examples/xsim.blanks.pdf",
  "doc/latex/xsim/examples/xsim.blanks.tex",
  "doc/latex/xsim/examples/xsim.boxed-headings.pdf",
  "doc/latex/xsim/examples/xsim.boxed-headings.tex",
  "doc/latex/xsim/examples/xsim.code-and-output.pdf",
  "doc/latex/xsim/examples/xsim.code-and-output.tex",
  "doc/latex/xsim/examples/xsim.collections.pdf",
  "doc/latex/xsim/examples/xsim.collections.tex",
  "doc/latex/xsim/examples/xsim.crossref.pdf",
  "doc/latex/xsim/examples/xsim.crossref.tex",
  "doc/latex/xsim/examples/xsim.description-list.pdf",
  "doc/latex/xsim/examples/xsim.description-list.tex",
  "doc/latex/xsim/examples/xsim.different-point-types.pdf",
  "doc/latex/xsim/examples/xsim.different-point-types.tex",
  "doc/latex/xsim/examples/xsim.difficulties.pdf",
  "doc/latex/xsim/examples/xsim.difficulties.tex",
  "doc/latex/xsim/examples/xsim.floating.pdf",
  "doc/latex/xsim/examples/xsim.floating.tex",
  "doc/latex/xsim/examples/xsim.golatex-80640.pdf",
  "doc/latex/xsim/examples/xsim.golatex-80640.tex",
  "doc/latex/xsim/examples/xsim.golatex-91339.pdf",
  "doc/latex/xsim/examples/xsim.golatex-91339.tex",
  "doc/latex/xsim/examples/xsim.grade-distribution.pdf",
  "doc/latex/xsim/examples/xsim.grade-distribution.tex",
  "doc/latex/xsim/examples/xsim.hints.pdf",
  "doc/latex/xsim/examples/xsim.hints.tex",
  "doc/latex/xsim/examples/xsim.listings.pdf",
  "doc/latex/xsim/examples/xsim.listings.tex",
  "doc/latex/xsim/examples/xsim.listofexercises.pdf",
  "doc/latex/xsim/examples/xsim.listofexercises.tex",
  "doc/latex/xsim/examples/xsim.multiplechoice.pdf",
  "doc/latex/xsim/examples/xsim.multiplechoice.tex",
  "doc/latex/xsim/examples/xsim.pointsums.pdf",
  "doc/latex/xsim/examples/xsim.pointsums.tex",
  "doc/latex/xsim/examples/xsim.randomexercises.pdf",
  "doc/latex/xsim/examples/xsim.randomexercises.tex",
  "doc/latex/xsim/examples/xsim.texsx-13635.pdf",
  "doc/latex/xsim/examples/xsim.texsx-13635.tex",
  "doc/latex/xsim/examples/xsim.texsx-155630.pdf",
  "doc/latex/xsim/examples/xsim.texsx-155630.tex",
  "doc/latex/xsim/examples/xsim.texsx-199360.pdf",
  "doc/latex/xsim/examples/xsim.texsx-199360.tex",
  "doc/latex/xsim/examples/xsim.texsx-299534.pdf",
  "doc/latex/xsim/examples/xsim.texsx-299534.tex",
  "doc/latex/xsim/examples/xsim.texsx-305110.pdf",
  "doc/latex/xsim/examples/xsim.texsx-305110.tex",
  "doc/latex/xsim/examples/xsim.texsx-308883.pdf",
  "doc/latex/xsim/examples/xsim.texsx-308883.tex",
  "doc/latex/xsim/examples/xsim.texsx-338165.pdf",
  "doc/latex/xsim/examples/xsim.texsx-338165.tex",
  "doc/latex/xsim/examples/xsim.texsx-350028.pdf",
  "doc/latex/xsim/examples/xsim.texsx-350028.tex",
  "doc/latex/xsim/examples/xsim.texsx-369065.pdf",
  "doc/latex/xsim/examples/xsim.texsx-369065.tex",
  "doc/latex/xsim/examples/xsim.texsx-369636.pdf",
  "doc/latex/xsim/examples/xsim.texsx-369636.tex",
  "doc/latex/xsim/examples/xsim.texsx-369803.pdf",
  "doc/latex/xsim/examples/xsim.texsx-369803.tex",
  "doc/latex/xsim/examples/xsim.texsx-370642.pdf",
  "doc/latex/xsim/examples/xsim.texsx-370642.tex",
  "doc/latex/xsim/examples/xsim.texsx-391530.pdf",
  "doc/latex/xsim/examples/xsim.texsx-391530.tex",
  "doc/latex/xsim/examples/xsim.texsx-395273.pdf",
  "doc/latex/xsim/examples/xsim.texsx-395273.tex",
  "doc/latex/xsim/examples/xsim.texsx-498299.pdf",
  "doc/latex/xsim/examples/xsim.texsx-498299.tex",
  "doc/latex/xsim/examples/xsim.texwelt-15093.pdf",
  "doc/latex/xsim/examples/xsim.texwelt-15093.tex",
  "doc/latex/xsim/examples/xsim.texwelt-23968.pdf",
  "doc/latex/xsim/examples/xsim.texwelt-23968.tex",
  "doc/latex/xsim/examples/xsim.texwelt-6698.pdf",
  "doc/latex/xsim/examples/xsim.texwelt-6698.tex",
  "doc/latex/xsim/examples/xsim.various.pdf",
  "doc/latex/xsim/examples/xsim.various.tex",
  "tex/latex/xsim/xsim.auxfile.code.tex",
  "tex/latex/xsim/xsim.base.code.tex",
  "tex/latex/xsim/xsim.blanks.code.tex",
  "tex/latex/xsim/xsim.collections.code.tex",
  "tex/latex/xsim/xsim.definitions.code.tex",
  "tex/latex/xsim/xsim.environments.code.tex",
  "tex/latex/xsim/xsim.exercises.code.tex",
  "tex/latex/xsim/xsim.goals.code.tex",
  "tex/latex/xsim/xsim.grades.code.tex",
  "tex/latex/xsim/xsim.interface.code.tex",
  "tex/latex/xsim/xsim-manual.cls",
  "tex/latex/xsim/xsim.modules.code.tex",
  "tex/latex/xsim/xsim.properties.code.tex",
  "tex/latex/xsim/xsim.random.code.tex",
  "tex/latex/xsim/xsim.solutions.code.tex",
  "tex/latex/xsim/xsim.sty",
  "tex/latex/xsim/xsim.style.layouts.code.tex",
  "tex/latex/xsim/xsim.tags.code.tex",
  "tex/latex/xsim/xsim.templates.code.tex",
  "tex/latex/xsim/xsim.translations.code.tex",
  "tex/latex/xsim/xsimverb.sty",
  "tex/latex/xsim/xsim.verbwrite.code.tex"
}

samples = {
  "doc/examples/xsim.blanks.tex",
  "doc/examples/xsim.boxed-headings.tex",
  "doc/examples/xsim.code-and-output.tex",
  "doc/examples/xsim.collections.tex",
  "doc/examples/xsim.crossref.tex",
  "doc/examples/xsim.description-list.tex",
  "doc/examples/xsim.different-point-types.tex",
  "doc/examples/xsim.difficulties.tex",
  "doc/examples/xsim.floating.tex",
  "doc/examples/xsim.golatex-80640.tex",
  "doc/examples/xsim.golatex-91339.tex",
  "doc/examples/xsim.grade-distribution.tex",
  "doc/examples/xsim.hints.tex",
  "doc/examples/xsim.listings.tex",
  "doc/examples/xsim.listofexercises.tex",
  "doc/examples/xsim.multiplechoice.tex",
  "doc/examples/xsim.pointsums.tex",
  "doc/examples/xsim.randomexercises.tex",
  "doc/examples/xsim.texsx-13635.tex",
  "doc/examples/xsim.texsx-155630.tex",
  "doc/examples/xsim.texsx-199360.tex",
  "doc/examples/xsim.texsx-299534.tex",
  "doc/examples/xsim.texsx-305110.tex",
  "doc/examples/xsim.texsx-308883.tex",
  "doc/examples/xsim.texsx-338165.tex",
  "doc/examples/xsim.texsx-350028.tex",
  "doc/examples/xsim.texsx-369065.tex",
  "doc/examples/xsim.texsx-369636.tex",
  "doc/examples/xsim.texsx-369803.tex",
  "doc/examples/xsim.texsx-370642.tex",
  "doc/examples/xsim.texsx-391530.tex",
  "doc/examples/xsim.texsx-395273.tex",
  "doc/examples/xsim.texsx-498299.tex",
  "doc/examples/xsim.texwelt-15093.tex",
  "doc/examples/xsim.texwelt-23968.tex",
  "doc/examples/xsim.texwelt-6698.tex",
  "doc/examples/xsim.various.tex",
}

-- Update package date and version
tagfiles = {"README", "xsim.sty", "xsimverb.sty"}
function update_tag(file, content, tagname, tagdate)
  if string.match(file, "README") then
    content = string.gsub(content,
                          "XSIM     %d%d%d%d%/%d%d%/%d%d v(%S+)",
                          "XSIM     "..xsimpkgd.." v"..xsimpkgv)
    content = string.gsub(content,
                          "XSIMVERB %d%d%d%d%/%d%d%/%d%d v(%S+)",
                          "XSIMVERB "..xsimverbd.." v"..xsimverbv)
  end
  if string.match(file, "xsim.sty") then
    local v_maj, v_min, v_sub = string.match(xsimpkgv, "^v?(%d+).?(%d+)(%S*)$")
    content = string.gsub(content,
                          "\\c_xsim_date_tl                 {%d%d%d%d%/%d%d%/%d%d}",
                          "\\c_xsim_date_tl                 {"..xsimpkgd.."}")
    content = string.gsub(content,
                          "\\c_xsim_version_major_number_tl {.-}",
                          "\\c_xsim_version_major_number_tl {"..v_maj.."}")
    content = string.gsub(content,
                          "\\c_xsim_version_minor_number_tl {.-}",
                          "\\c_xsim_version_minor_number_tl {"..v_min.."}")
    content = string.gsub(content,
                          "\\c_xsim_version_subrelease_tl   {.-}",
                          "\\c_xsim_version_subrelease_tl   {"..v_sub.."}")
  end
  if string.match(file, "xsimverb.sty") then
    local v_maj, v_min, v_sub = string.match(xsimverbv, "^v?(%d+).?(%d*)(%S*)$")
    content = string.gsub(content,
                          "\\c_xsimverb_date_tl                 {%d%d%d%d%/%d%d%/%d%d}",
                          "\\c_xsimverb_date_tl                 {"..xsimverbd.."}")
    content = string.gsub(content,
                          "\\c_xsimverb_version_major_number_tl {.-}",
                          "\\c_xsimverb_version_major_number_tl {"..v_maj.."}")
    content = string.gsub(content,
                          "\\c_xsimverb_version_minor_number_tl {.-}",
                          "\\c_xsimverb_version_minor_number_tl {"..v_min.."}")
    content = string.gsub(content,
                          "\\c_xsimverb_version_subrelease_tl   {.-}",
                          "\\c_xsimverb_version_subrelease_tl   {"..v_sub.."}")
  end
  return content
end


-- Build xsim_manual.tex using arara
typesetfiles = {"xsim_manual.tex"}

local function type_manual()
  errorlevel = (mkdir(typesetdir.."/exercises")
              + mkdir(typesetdir.."/examples"))
  if errorlevel ~= 0 then
    error("** Error!!: The ./exercises and ./examples directory could not be created")
    return errorlevel
  end
  for i, samples in ipairs(samples) do
    local file = jobname(samples)
    errorlevel = (cp(file..".tex", "doc/examples", typesetdir.."/examples")
                + cp(file..".pdf", "doc/examples", typesetdir.."/examples"))
    if errorlevel ~= 0 then
      error("** Error!!: Can't copy "..file.." from ./doc/examples to ./"..typesetdir)
    return errorlevel
    end
  end
  errorlevel = cp("libertinehologopatch.sty", "doc", typesetdir)
  errorlevel = cp("*.cls", "doc", typesetdir)
  if errorlevel ~= 0 then
    error("** Error!!: Can't copy libertinehologopatch.sty from ./doc to ./"..typesetdir)
    return errorlevel
  end
  local file = jobname("doc/xsim_manual.tex")
  errorlevel = runcmd("arara "..file..".tex", typesetdir, {"TEXINPUTS","LUAINPUTS"})
  if errorlevel ~= 0 then
    error("Error!!: Typesetting "..file..".tex")
    return errorlevel
  end
  return 0
end

specialtypesetting = { }
specialtypesetting["xsim_manual.tex"] = {func = type_manual}

-- A new target "examples" to run the examples files
if options["target"] == "examples" then
  os.execute("l3build unpack -q")
  for i, samples in ipairs(samples) do
    local file = jobname(samples)
    errorlevel = cp(file..".tex", "doc/examples", "build/unpacked")
    if errorlevel ~= 0 then
      error("** Error!!: Can't copy "..file..".tex from ./doc/examples to ./build/unpacked")
    return errorlevel
    end
    errorlevel = run("build/unpacked", "pdflatex "..file..".tex")
    if errorlevel ~= 0 then
      error("** Error!!: pdflatex "..file..".tex")
      return errorlevel
    else
      print("** Running: pdflatex "..file..".tex")
    end
  end
  os.exit(0)
end
